### ISSUE TITLE
Using dask with lots of permutations and large block size leads to high memory usage

### ISSUE THREAD

**Issue by mprignon** [girpas-ulg/trendfit#2]
Using Dask LocalCluster (Distributed) with >10k samples and large block size (>1000) leads to very high memory usage (>10GB). It's thus almost impossible to use several Dask workers.  
From https://docs.dask.org/en/latest/delayed-best-practices.html, we can solve this by breaking up the delayed list into smaller batches.  
In _bootsrap.py starting at line 47:
```
if OPTIONS['use_dask']:
            import dask
            import dask.bag as db

            sequence = [np.random.RandomState()
                        for _ in range(self.n_samples)]
            b = db.from_sequence(sequence, npartitions=100)
            dlyd = b.map(fit_sample).to_delayed()

            res = dask.compute(*dlyd) # returns a list of 100 lists (npartitions=100) of results (because of dask bag) 
            res = [item for sublist in res for item in sublist] # flats the list of lists
           """
           original code:
            dlyd = [dask.delayed(fit_sample)(np.random.RandomState())
                    for _ in range(self.n_samples)]            
           res = dask.compute(*dlyd)
           """


```

This modification seems to lead to lower memory usage, keeping the same performance.

**Comment by benbovy** [comment]
Yes, I think that's a good workaround to avoid too many tasks being send to Dask's scheduler. Would you like to prepare and submit a pull-request with the change that you suggest? I'm happy to review it!

Regarding memory usage, I think it may rather be a side effect. I guess you use `save_models=False` for such a large number of bootstrap samples? Even when not saving the models, it might be possible that the memory is not released for some reason. Trying to force `del mb` within `fit_sample()` might help.

**Comment by mprignon** [comment]
Thanks for the answer !

> I guess you use save_models=False for such a large number of bootstrap samples?


Yes

I have submitted a pull-request (first time I do this so I hope everything is ok) with the suggested changes.

**Comment by benbovy** [comment]
Closed in #3.
